aea/manager.py

```diff
@@ -0,0 +1,534 @@
+                task.wait(): agent_name
+                for agent_name, task in self._agents_tasks.items()
+            }
+            wait_tasks = list(tasks_for_agents.keys()) + [self._event.wait()]  # type: ignore
+            done, _ = await asyncio.wait(wait_tasks, return_when=FIRST_COMPLETED)
+
+            if self._event.is_set():
+                self._event.clear()
+
+            for task in done:
+                if task not in tasks_for_agents:
+                    await task
+                    continue
+                agent_name = tasks_for_agents[task]
+                self._agents_tasks.pop(agent_name)
+                if task.exception():
+                    print(agent_name, "exceptioned", task.exception())
+                    for callback in self._error_callbacks:
+                        callback(agent_name, task.exception())
+                else:
+                    await task
+
+    def start_manager(self) -> None:
+        """Start manager."""
+        self._ensure_working_dir()
+        self._started_event.clear()
+        self._is_running = True
+        self._thread.start()
+        self._started_event.wait(self.DEFALUT_TIMEOUT_FOR_BLOCKING_OPERATIONS)
+
+    def stop_manager(self, cleanup: bool = False) -> None:
+        """
+        Stop manager.
+
+        Stops all running agents and stop agent.
+
+        :param cleanup: remove agents_dir and purge in memory registry.
+
+        :return: None
+        """
+        if not self._loop or not self._event:
+            raise ValueError("Manager was not started!")
+        if self._was_working_dir_created:
+            rmtree(self.working_dir)
+
+        if self._thread.is_alive():
+            self.stop_all_agents()
+        self._is_running = False
+        self._loop.call_soon_threadsafe(self._event.set)
+        self._thread.join(self.DEFALUT_TIMEOUT_FOR_BLOCKING_OPERATIONS)
+
+    def add_project(self, public_id: PublicId) -> None:
+        """Fetch agent project and all dependencies to working_dir."""
+        if public_id in self._projects:
+            raise ValueError(f"Project {public_id} was already added!")
+        self._projects[public_id] = Project.load(self.working_dir, public_id)
+
+    def remove_project(self, public_id: PublicId) -> None:
+        """Remove agent project."""
+        if public_id not in self._projects:
+            raise ValueError(f"Project {public_id} was not added!")
+
+        if self._projects[public_id].agents:
+            raise ValueError(
+                f"Can not remove projects with aliases exists: {self._projects[public_id].agents}"
+            )
+
+        self._projects.pop(public_id).remove()
+
+    def list_projects(self) -> List[PublicId]:
+        """
+        List all agents projects added.
+
+        :return: lit of public ids of projects
+        """
+        return list(self._projects.keys())
+
+    def add_agent(
+        self,
+        public_id: PublicId,
+        agent_name: str,
```

Comment:
Should be optional the agent_name. If I add only one agent instance for a given public_id, then the name from the public id should be taken.

Reply:
not the most obvious i suppose.
```
manager.add_agent(PublicID1)
agent2_name = 'name2'
manager.add_agent(PublicID1, name=agent2_name)

manager.start_agent(PublicID1.name)
manager.start_agent(agent2_name)
```